src/local.c

```diff
@@ -18,33 +18,249 @@
 #include <stdio.h>
 #include <string.h>
 #include <unistd.h>
+#include <utmp.h>
+#include <stdlib.h>
+#include <dirent.h>
+#include <fcntl.h>
 #include "log.h"
 #include "conf.h"
 #include "process.h"
+#include "tmux.h"
 
-int pusb_local_login(t_pusb_options *opts, const char *user)
+int pusb_is_tty_local(char *tty)
+{
+	struct utmp	utsearch;
+	struct utmp	*utent;
+
+	if (strstr(tty, "/dev/") != NULL) {
+		tty += strlen("/dev/");
+	}
+
+	strncpy(utsearch.ut_line, tty, sizeof(utsearch.ut_line) - 1);
+
+	setutent();
+	utent = getutline(&utsearch);
+	endutent();
+
+	if (!utent)
+	{
+		log_debug("	No utmp entry found for tty \"%s\"\n", utsearch.ut_line);
+		return (0);
+	} else {
+		log_debug("		utmp entry for tty \"%s\" found\n", tty);
+		log_debug("			utmp->ut_pid: %d\n", utent->ut_pid);
+		log_debug("			utmp->ut_user: %s\n", utent->ut_user);
+	}
+
+	for (int i = 0; i < 4; ++i)
+	{
+		/**
+		 * Note: despite the property name this also works for IPv4, v4 addr would be in ut_addr_v6[0] solely.
+		 * See utmp man (http://manpages.ubuntu.com/manpages/bionic/de/man5/utmp.5.html)
+		 **/
+		if (utent->ut_addr_v6[i] != 0)
+		{
+			log_error("Remote authentication request: %s\n", utent->ut_host);
+			return (-1);
+		} else {
+			log_debug("		Checking utmp->ut_addr_v6[%d]\n", i);
+		}
+	}
+
+	log_debug("	utmp check successful, request originates from a local source!\n");
+	return (1);
+}
+
+char *pusb_get_tty_from_xorg_process(const char *display)
+{
+	DIR *d_proc = opendir("/proc");
+	if (d_proc == NULL)
+		return NULL;
+
+	char *expected_core = (char *)malloc(12);
+	char *cmdline_path = (char *)malloc(32);
+	char *cmdline = (char *)malloc(4096);
+	char *fd_path = (char *)malloc(32);
+	char *link_path = (char *)malloc(32);
+	char *fd_target = (char *)malloc(32);
+
+	sprintf(expected_core, "-core %s", display);
+
+	struct dirent *dent_proc;
+	while ((dent_proc = readdir(d_proc)) != NULL)
+	{
+		if (dent_proc->d_type == DT_DIR && atoi(dent_proc->d_name) != 0 && strcmp(dent_proc->d_name, ".") != 0 && strcmp(dent_proc->d_name, "..") != 0)
+		{
+			memset(cmdline_path, 0, 32);
+			sprintf(cmdline_path, "/proc/%s/cmdline", dent_proc->d_name);
+
+			memset(cmdline, 0, 4096);
+			int cmdline_file = open(cmdline_path, O_RDONLY | O_CLOEXEC);
```

Comment:
refactor to fopen/fread/fclose to keep it uniform

Reply:
will stay like this for now. this is tested, and the `\0`s in the file being read could maybe cause issues.